prometheus/metrics.go

```diff
@@ -58,17 +58,21 @@ func buildLabelStrings(q *IstioMetricsQuery) (string, []string) {
 
 	errors := []string{}
 	protocol := strings.ToLower(q.RequestProtocol)
-	if protocol == "" || protocol == "grpc" {
-		// this is intentionally not `grpc_response_status!="0"`. We need to be backward compatible and
-		// handle the case where grpc_response_status does not exist.  In Prometheus, negative tests on a
-		// non-existent label match everything, but positive tests match nothing. So, we stay positive.
-		grpcLabels := append(labels, `grpc_response_status=~"^[1-9]$|^1[0-6]$"`)
+
+	// both http and grpc requests can suffer from no response (response_code=0) or an http error
+	// (response_code=4xx,5xx), and so we always perform a query against response_code:
+	httpLabels := append(labels, `response_code=~"^0$|^[4-5]\\d\\d$"`)
+	errors = append(errors, "{"+strings.Join(httpLabels, ",")+"}")
+
+	// if necessary also look for grpc errors. note that the grpc test intentionally avoids
+	// `grpc_response_status!="0"`. We need to be backward compatible and handle the case where
+	// grpc_response_status does not exist, or if it is simply unset. In Prometheus, negative tests on a
+	// non-existent label match everything, but positive tests match nothing. So, we stay positive.
+	// furthermore, make sure we only count grpc errors with successful http status.
+	if protocol != "http" {
+		grpcLabels := append(labels, `grpc_response_status=~"^[1-9]$|^1[0-6]$",response_code!~"^0$|^[4-5]\\d\\d$"`)
```

Comment:
since that exact same regex is used twice, here and above, perhaps put it as a const to make sure one isn't modified without the other